### ISSUE TITLE
[Bug] (B2W2) Driftveil Cup mostly unaffected by Trainer level increase

### ISSUE THREAD

**Issue by FrostedGeulleisia** [Ajarmar/universal-pokemon-randomizer-zx#443]
**Describe the bug**
When the Trainer Pokémon level increase is applied in B2W2, it ends up not affecting the mandatory Driftveil cup segment 5 badges into the game's storyline.
Note that it's correct in the spoiler log, which makes this bug even more mysterious to me, as in-game the levels are all 25 flat:
`#368 (\[PK]\[MN] Trainer Rival) - Castform@Enigma Berry Lv50, Espeon@Starf Berry Lv47, Tangrowth@Aspear Berry Lv47, Grotle@Expert Belt Lv47, Aggron@Razor Claw Lv47`
Also of note is that three random Pokémon are chosen from ~~this team~~ (the other two opponent's teams), that however I can imagine as intended behavior.
EDIT: Above is true only for Cheren and Colress, for the Rival, all of them appeared, at least on a different seed.
Perhaps the game automatically scales the opponent's levels manually? Maybe it should at least be noted in the hover text for the option? Why would they scale only the opponent if their levels are always going to be the same? Why don't scale the player as well? There are entirely too many questions for me to list them all here, but maybe it has something to do with how the PWT works *after* the initial Driftveil cup fights.

What's also interesting, is that neither the Driftveil cup Cheren nor Colress fights appear in the spoiler log. All three Rival rosters do, however.

**To Reproduce**
1. Randomize with any settings,  but apply a significant enough trainer level increase
2. Go to the Driftveil Cup in the storyline
3. All opponents will be level 25 flat.

**Expected behavior**
I expected the levels of opponents in the Driftveil cup to be as advertised in the spoiler log: (47-50 with 50% increase), instead of what's shown in-game, and obviously, only a party size of 3.

**Screenshots**
![image](https://user-images.githubusercontent.com/27410260/169853366-0f18c7f9-d78b-43a8-80cd-0017d3d0bc05.png)
The rest of the team matches the snippet from the spoiler log provided earlier, barring the levels.

**System (please complete the following information):**
 - Operating System: Win10 Home x64 (21H1)
 - Java Version: 8 update 291 (when i tried to update it just kinda, uninstalled the old version but didn't install the new one but the rando still works? no idea at this point)
 - Randomizer Version: v4.4.0 (tried on latest master as well)

**Additional context**
Example here provided with B2, a seed of 62333825394152 and a settings string of `318QQgEBgLuAfEB9AGRAAQeoWEECQEACQANAR8iAAAAAAAABAKI5FnkwScJCTIGBgDkFwETUG9rZW1vbiBCbGFjayAyIChVKQhPirXjwziK`
We have however confirmed it occurs independently of version.
A friend's save file in front of the segment in question is also attached (in a zip, because Github doesn't like the extension otherwise), it's from a different seed, but still works for our purposes.
[test.zip](https://github.com/Ajarmar/universal-pokemon-randomizer-zx/files/8755973/test.zip)

I honestly think that, if this cap can't be patched out, it should at least be noted in the randomizer itself that said cap exists.

**Comment by FrostedGeulleisia** [comment]
Also a little, slightly unrelated question, are bugs encountered on master but unable to be reproduced with the current stable release due to the inability to import the seed/settings valid? I assume no, but please let me know in case they are.

**Comment by tom-overton** [comment]
The Driftveil cup stuff is implemented in a frankly insane manner, so some of your confusion is just "that's how the game works". Let me step through things one-by-one:

**Why does Hugh appear in the logs with scaled levels and more Pokemon?**
This is because the Hugh you're looking at in the logs is actually the Hugh that helps you in a Multi Battle on the Plasma Frigate, which happens immediately after you clear the Driftveil Cup. For whatever reason, the game uses *this* team to generate his Driftveil Cup team; he doesn't have a distinct trainer entry for the PWT like the other trainers. This is also why, if you give Hugh additional Pokemon, he will use more than three Pokemon in the Driftveil Cup. He isn't bound by the same "rules" as other PWT Trainers because his team is generated in a different manner.

**Why don't Colress/Cheren show up in the logs?**
They actually *do* show up, but they aren't labeled as Cheren and Colress, or really anything at all. We weren't reading the PWT names and class names in, so those trainers only had their Pokemon appear, not their names. This was fixed in 0fd1898d944f328bf0061297d7e5fe904dce26b6, so it will be present in the next release. Here's an example from a log I just made using the latest `master` commit:
```
#814 (\[PK]\[MN] Trainer Cheren => Artisan Cristiano) - Stoutland Lv25, Cinccino Lv25, Watchog Lv25
#815 (\[PK]\[MN] Trainer Colress => Artisan Scott) - Klink Lv25, Elgyem Lv25, Magneton Lv25
```

**Why are enemy trainers not affected by the level scaling?**
The Driftveil Cup is handled differently from other trainers. Unlike "normal" trainer data, the data for PWT trainers doesn't contain any sort of level. How the level gets determined is, at this point, unknown, but it's fixed at level 25. Therefore, the randomizer just hardcodes a level of 25 for these trainers, but we don't actually read from or write to it anywhere.

**Why isn't the player scaled to level 25 during the Driftveil Cup?**
This is a question for the folks at Game Freak Co., Ltd., not for the randomizer.  We don't change how this works at all.

**Why doesn't the hover text state that the Driftveil Cup Pokemon are unaffected by the level increase?**
We just didn't think of it at the time. I can talk to Ajarmar about it, but we probably won't change it, because we usually don't call out extremely game-specific things in the tooltips.

Now, about the bug in the title itself: we don't currently know *how* the game forces these Pokemon to be level 25, so that means we also don't know how to change it. Additionally, Ajarmar brought up that these trainers grant no experience, so we might not *want* to scale their levels; it would be a massive pain to be stuck on these trainers and not gain any EXP from them. So I'm going to close this bug for now, since we have no idea how to change this, nor any plan to do the research on how to change it. Hopefully, the above explanations make the craziness of the Driftveil Cup a little more understandable.

Lastly, if you have a bug that you have seen on the latest `master` commit, please feel free to file an issue! As long as you use the latest commit and attach a log file, any issue should be reproducible.
